stan/math/prim/fun/select.hpp

```diff
@@ -0,0 +1,180 @@
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = return_type_t<T_true, T_false>,
+          require_all_stan_scalar_t<T_true, T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true y_true,
+                      const T_false y_false) {
+  return c ? ReturnT(y_true) : ReturnT(y_false);
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument. Eigen expressions are
+ * evaluated so that the return type is the same for both branches.
+ *
+ * Both containers must have the same plain type. The scalar type
+ * of the return is determined by the return_type_t<> type trait.
+ *
+ * Overload for use with two containers.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <
+    typename T_true, typename T_false,
+    typename T_return = return_type_t<T_true, T_false>,
+    typename T_true_plain = promote_scalar_t<T_return, plain_type_t<T_true>>,
+    typename T_false_plain = promote_scalar_t<T_return, plain_type_t<T_false>>,
+    require_all_container_t<T_true, T_false>* = nullptr,
+    require_all_same_t<T_true_plain, T_false_plain>* = nullptr>
+inline T_true_plain select(const bool c, T_true&& y_true, T_false&& y_false) {
+  check_matching_dims("select", "left hand side", y_true, "right hand side",
+                      y_false);
+  return c ? T_true_plain(std::forward<T_true>(y_true))
+           : T_true_plain(std::forward<T_false>(y_false));
+}
+
+/**
+ * Return the second argument if the first argument is true
+ * and otherwise return the third argument.
+ *
+ * Overload for use when the 'true' return is a container and the 'false'
+ * return is a scalar
+ *
+ * If chosen, the scalar is returned as a container of the same size and
+ * plain type as the provided argument. Consequently, any Eigen expressions are
+ * evaluated.
+ *
+ * @tparam T_true type of the true argument
+ * @tparam T_false type of the false argument
+ * @param c Boolean condition value.
+ * @param y_true Value to return if condition is true.
+ * @param y_false Value to return if condition is false.
+ */
+template <typename T_true, typename T_false,
+          typename ReturnT = promote_scalar_t<return_type_t<T_true, T_false>,
+                                              plain_type_t<T_true>>,
+          require_container_t<T_true>* = nullptr,
+          require_stan_scalar_t<T_false>* = nullptr>
+inline ReturnT select(const bool c, const T_true& y_true,
+                      const T_false& y_false) {
+  if (c) {
+    return y_true;
+  } else {
+    return apply_scalar_binary(
+        y_true, y_false,
+        [](const auto& y_true_inner, const auto& y_false_inner) {
+          return y_false_inner;
+        });
```

Comment:
What is the use case for the container and scalar overload? Is this for like broadcasting? imo I'd rather do this sort of conditional at compile time and return either a scalar or a container like `as_array_or_scalar()`

Reply:
My worry here is that folks are going to accidentally use this overload and get back a container when they instead wanted a scalar.

Reply:
Yep the broadcasting is the intended outcome (and is doc-ed). 

See [this line from the Bernoulli PR](https://github.com/stan-dev/math/pull/2784/files#diff-1aafd077ec7f8966bf7f92be2addaee531c6be68b53a5e798681c8c279e60124R63) for an example of usage where the inputs can be scalars or containers, and the return needs to be broadcast (if necessary) to the common container type & size